dao-voting/src/main.sw

```diff
@@ -0,0 +1,295 @@
+        let approval_percentage = proposal.yes_votes * 100 / (proposal.yes_votes + proposal.no_votes);
+        require(approval_percentage >= proposal.approval_percentage, Error::ApprovalPercentageNotMet);
+
+        asm(rA: proposal.call_data.memory_address, rB: proposal.call_data.num_coins_to_forward, rC: proposal.call_data.asset_id_of_coins_to_forward, rD: proposal.call_data.amount_of_gas_to_forward) {
+            call rA rB rC rD;
+        }
+
+        // Users can now convert their votes back into tokens
+
+        true
+    }
+
+    /// Unlock tokens used to vote on proposals to allow the user to withdraw
+    /// If the user had not voted in the given expired proposal, nothing happens
+    ///
+    /// # Parameters
+    /// proposal_id - proposal to turn user votes back into governance tokens
+    ///
+    /// # Panics
+    ///
+    /// This function will panic when:
+    /// - The constructor has not ben called to initialize
+    /// - The proposal id is invalid
+    /// - The proposal is still active
+    #[storage(read, write)]fn convert_votes_to_tokens(proposal_id: u64) {
+        require(storage.state == 1, Error::NotInitialized);
+        require(proposal_id < storage.proposal_count, Error::InvalidId);
+
+        let proposal = storage.proposals.get(proposal_id);
+        require(proposal.end_height < height(), Error::ProposalActive);
+
+        let sender: Identity = sender_identity();
+        let votes = storage.votes.get((sender, proposal_id));
+
+        storage.votes.insert((sender, proposal_id), 0);
+
+        let balance = storage.balances.get(sender);
+        storage.balances.insert(sender, balance + votes);
+    }
+
+    /// Return the amount of governance tokens in this contract
+    #[storage(read)]fn balance() -> u64 {
+        this_balance(storage.gov_token)
+    }
+
+    /// Return the amount of governance tokens a user has in this contract
+    ///
+    /// # Parameters
+    /// user - user of which to get internal balance of governance tokens
+    #[storage(read)]fn user_balance(user: Identity) -> u64 {
+        storage.balances.get(user)
+    }
+
+    /// Return the amount of votes a user has used on a proposal
+    ///
+    /// # Parameters
+    /// user - user of which to get votes spent on a proposal
+    /// proposal_id - proposal of which to get votes spent by user
+    #[storage(read)]fn user_votes(user: Identity, proposal_id: u64) -> u64 {
+        require(proposal_id < storage.proposal_count, Error::InvalidId);
+        storage.votes.get((user, proposal_id))
+    }
+
+    /// Return proposal data for a given id
+    ///
+    /// # Parameters
+    /// proposal_id - id of proposal to get
+    ///
+    /// # Panics
+    ///
+    /// The function will panic when:
+    /// - The constructor has not been called ot initialize
+    /// - The given id is out of range
+    #[storage(read)]fn proposal(proposal_id: u64) -> Proposal {
+        require(storage.state == 1, Error::NotInitialized);
```

Comment:
Why do you need to revert here if not initialized? I would remove

Reply:
also, you've not handled the id 0 case

Reply:
I think this should also be <= instead of <

Reply:
What do you mean the id 0 case?  Why can't a proposal_id be 0.

Reply:
Currently I'm indexing the proposals in the storage map how I would with a vector/array, starting at the 0 index.

Reply:
If an incorrect query is made then 0 could be used as a sentinel to indicate that there is no data there. It is clearer than returning actual data and then having to notice that the data that you are expecting is not the one that you have received. 
It's fine if you want to continue to use 0 though